internal/services/stock_exchange/algorithms.go

```diff
@@ -0,0 +1,340 @@
+package stock_exchange
+
+import (
+	"Tradeasy/config"
+	model "Tradeasy/internal/model/stock_exchange"
+	"fmt"
+	"math"
+	"time"
+)
+
+type OrderId struct {
+	id int
+}
+
+var orderResponse = make(chan OrderResponse)
+
+func UpdateLTP(ltp int, stock string) {
+	currentStock := model.Stocks{}
+	config.DB.Raw("SELECT * FROM stocks WHERE stock_name = ?", stock).Scan(&currentStock)
+	currentStock.HighPrice = int(math.Max(float64(ltp), float64(currentStock.HighPrice)))
+	currentStock.LowPrice = int(math.Min(float64(ltp), float64(currentStock.LowPrice)))
+	currentStock.LTP = ltp
+	currentStock.UpdatedAt = time.Now()
+	//fmt.Println(currentStock)
+	config.DB.Exec("UPDATE stocks SET ltp = ? ,high_price = ? ,low_price = ? , updated_at = ?  WHERE stock_ticker_symbol = ? ", ltp, currentStock.HighPrice, currentStock.LowPrice, currentStock.UpdatedAt, stock)
+	//config.DB.Table("stocks").Save(&currentStock)
+}
+
+func BuyOrderMatchingAlgo(buyOrderBody OrderRequest, resp OrderResponse, id int) {
+
+	var sellBook []model.SellOrderBook
+	// db lock
+	err := config.DB.Raw("SELECT * FROM sell_order_book WHERE stock_ticker_symbol = ?  ORDER BY order_price ASC,created_at ASC ", buyOrderBody.StockName).Scan(&sellBook).Error
+	if err != nil {
+		resp.Status = "Db Fetch failed, order cancelled"
+		config.DB.Exec("DELETE FROM buy_order_book WHERE id = ?", id)
+		orderResponse <- resp
+		return
+	}
+	if len(sellBook) == 0 {
+		// abort with message not enough shares
+		resp.Status = "Db Fetch failed, order cancelled"
+		orderResponse <- resp
+		config.DB.Exec("DELETE FROM buy_order_book WHERE id = ?", id)
+		return
+	}
+	SharesAvailable := 0
+	for _, elem := range sellBook {
+		SharesAvailable += elem.OrderQuantity
+	}
+
+	if SharesAvailable < buyOrderBody.Quantity {
```

Comment:
What about partial fulfillment?

Reply:
That makes the code very complicated. Not enough time.